File: apps/bridge-core/src/server/registry.ts
import type { BrowserConnection, BrowserStatus } from '@browser-bridge/shared';
import type { ServerWebSocket } from 'bun';
import type { WsData } from './types';

/**
 * Tracks inbound extension WebSocket connections, indexed by browserId.
 *
 * Pre-merge history: this class lived on the WebSocket Server and tracked
 * the upstream `ServerWebSocket<WsData>` instances that local-proxy dialed
 * into ws-server. After the bridge-core merge the equivalent of that handle
 * is the inbound `ServerWebSocket<undefined>` that the extension itself
 * dials into bridge-core on port 3002. The class name and surface stay;
 * what the tracked handle *points at* is now a different WebSocket. See
 * `docs/adr/0010-bridge-core-merges-ws-server-and-local-proxy.md`.
 */
interface RegistryEntry {
  browserId: string;
  userId: string;
  ws: ServerWebSocket<WsData>;
  status: BrowserStatus;
  lastSeen: number;
}

export class ConnectionRegistry {
  private browsers = new Map<string, RegistryEntry>();

  async register(
    ws: ServerWebSocket<WsData>,
    browserId: string,
  ): Promise<{ success: boolean; error?: string }> {
    const existing = this.browsers.get(browserId);
    if (existing && existing.ws !== ws && existing.ws.readyState === 1) {
      return { success: false, error: 'browser_id_in_use' };
    }

    const userId = ws.data.userId ?? 'unknown';

    this.browsers.set(browserId, {
      browserId,
      userId,
      ws,
      status: 'offline',
      lastSeen: Date.now(),
    });

    ws.data = { ...ws.data, browserId, userId };
    return { success: true };
  }

  setStatus(browserId: string, status: BrowserStatus): boolean {
    const entry = this.browsers.get(browserId);
    if (!entry) {
      // The pre-merge ws-server required an explicit `register` event before
      // any status could be set; the bridge-core merge dropped that
      // handshake — extensions connect directly to BrowserServer — so the
      // router now writes the entry lazily on the first status update.
      this.browsers.set(browserId, {
        browserId,
        userId: 'extension',
        // We do not have a reference to the extension WS here; the browser
        // connection lives in BrowserServer.extensionWs, not the registry.
        // Listing still works; getWebSocket() will return undefined.
        ws: null as unknown as ServerWebSocket<WsData>,
        status,
        lastSeen: Date.now(),
      });
      return true;
    }
    entry.status = status;
    entry.lastSeen = Date.now();
    return true;
  }

  getStatus(browserId: string): BrowserStatus | undefined {
    return this.browsers.get(browserId)?.status;
  }

  getWebSocket(browserId: string): ServerWebSocket<WsData> | undefined {
    return this.browsers.get(browserId)?.ws;
  }

  removeByWebSocket(ws: ServerWebSocket<WsData>): string | undefined {
    for (const [browserId, entry] of this.browsers) {
      if (entry.ws === ws) {
        this.browsers.delete(browserId);
        return browserId;
      }
    }
    return undefined;
  }

  listBrowsers(): BrowserConnection[] {
    return Array.from(this.browsers.values()).map(({ ws: _, ...rest }) => rest);
  }
}

File: apps/bridge-core/internal/core/registry.go
package core

import (
	"sync"
	"time"
)

// extensionUserID is the userId stamped on lazily created entries
// ('extension' in registry.ts).
const extensionUserID = "extension"

type entry struct {
	browserID string
	userID    string
	status    BrowserStatus
	lastSeen  int64
}

// Registry tracks per-browser status — the Go port of src/server/registry.ts:
// browserId-keyed connection bookkeeping. Post-merge entries are created
// lazily on the first status update (the extension connects straight to the
// browser server, so there is no register handshake); the tracked WebSocket
// handle from the pre-merge design is gone — the extension connection lives
// in the ws package (BrowserServer). Safe for concurrent use.
type Registry struct {
	mu      sync.Mutex
	entries map[string]*entry
	order   []string // insertion order, mirroring the TS Map iteration order
}

func NewRegistry() *Registry {
	return &Registry{entries: make(map[string]*entry)}
}

// SetStatus updates the entry for browserID, creating it lazily on first
// sight (registry.ts: the bridge-core merge dropped the explicit register
// handshake, so the router writes the entry on the first status update).
func (r *Registry) SetStatus(browserID string, status BrowserStatus) bool {
	r.mu.Lock()
	defer r.mu.Unlock()
	now := time.Now().UnixMilli()
	e, ok := r.entries[browserID]
	if !ok {
		r.entries[browserID] = &entry{
			browserID: browserID,
			userID:    extensionUserID,
			status:    status,
			lastSeen:  now,
		}
		r.order = append(r.order, browserID)
		return true
	}
	e.status = status
	e.lastSeen = now
	return true
}

// GetStatus returns the status and whether the registry has ever seen the
// browser (registry.ts returns undefined for unknown ids, and the inbound
// server rejects only those).
func (r *Registry) GetStatus(browserID string) (BrowserStatus, bool) {
	r.mu.Lock()
	defer r.mu.Unlock()
	e, ok := r.entries[browserID]
	if !ok {
		return "", false
	}
	return e.status, true
}

// ListBrowsers returns the list_browsers payload shape. The result is never
// nil so it marshals as [] rather than null.
func (r *Registry) ListBrowsers() []BrowserConnection {
	r.mu.Lock()
	defer r.mu.Unlock()
	out := make([]BrowserConnection, 0, len(r.entries))
	for _, id := range r.order {
		e := r.entries[id]
		out = append(out, BrowserConnection{
			BrowserID: e.browserID,
			UserID:    e.userID,
			Status:    e.status,
			LastSeen:  e.lastSeen,
		})
	}
	return out
}
